design/one-pager-pluggable-secret-stores.md

```diff
@@ -0,0 +1,269 @@
+kind: VaultConfig
+metadata:
+  name: vault-internal
+spec:
+  server: http://vault.vault-system:8200
+  mountPath: secrets/
+  version: v2
+  auth:
+    method: Token
+    token:
+      source: Secret
+      secretRef:
+        namespace: crossplane-system
+        name: vault-token
+        key: token
+```
+
+There won't be any API changes at MR/XR/XRC level as they would still refer to
+a `StoreConfig` resource as before.
+
+### Implementation
+
+We will implement the `External` [Secret Store] type in Crossplane Runtime which
+will use a gRPC client to communicate with the plugin server using the service
+definition below:
+
+```protobuf
+syntax = "proto3";
+
+// This service defines the APIs for an External Secret Store plugin.
+service ExternalSecretStoreService {
+    rpc GetSecret(GetSecretRequest) returns (GetSecretResponse) {}
+    rpc ApplySecret(ApplySecretRequest) returns (ApplySecretResponse) {}
+    rpc DeleteKeys(DeleteKeysRequest) returns (DeleteKeysResponse) {}
+}
+
+message ConfigReference {
+    string api_version = 1;
+    string kind = 2;
+    string name = 3;
+}
+message Secret {
+    string scoped_name = 1;
+    map<string, string> metadata = 2;
+    map<string, bytes> data = 3;
+}
+message GetSecretRequest {
+    ConfigReference config = 1;
+    Secret secret = 2;
+    GetOptions options = 3;
+}
+message GetSecretResponse {
+    Secret secret = 1;
+}
+message ApplySecretRequest {
+    ConfigReference config = 1;
+    Secret secret = 2;
+    ApplyOptions options = 3;
+}
+message ApplySecretResponse {
+    bool changed = 1;
+}
+message DeleteKeysRequest {
+    ConfigReference config = 1;
+    Secret secret = 2;
+    DeleteOptions options = 3;
+}
+message DeleteKeysResponse {}
+message GetOptions {}
+message ApplyOptions {}
+message DeleteOptions {
+    bool keep_empty_secret = 1;
+}
+```
+
+A _Secret Store Plugin_:
+
+- Will be a binary packaged as a container image.
+- Should run a gRPC server listening at Unix domain socket.
+- May support one or more Secret Stores.
+- Should deploy CRD(s) for supported store configurations at startup.
+- Will run as a sidecar next to Crossplane and Provider containers.
+
+#### Plugin Binary vs Secret Store Implementation
+
+All `rpc` methods have `ConfigReference` as their first parameter. So,
+one Secret Store Plugin may support more than one Secret Store implementation,
+e.g. Vault, AWS, GCP, Azure ... all in a single Plugin binary, and related
+implementation could be called depending on a given config similar to how it is
+being done in [RuntimeStoreBuilder] today. Implementation of individual Secret
+Stores still need to satisfy the same [Store] interface today and could live in
+separate repositories. However, they would need to be packaged and served by a
+Secret Store Plugin binary to be used with Crossplane.
+
+This approach allows us to support more than one Secret Store implementation
+with a single plugin sidecar simplifying the deployment model.
```

Comment:
I was imagining each secret store implemented and released separately supporting only one type.
Then, there could be another wrapper project(s), that imports existing secret stores and supports multiple. Not perfect but I feel like a good optimization to simplify the deployment model.

Reply:
Is there a reason we couldn't just deploy multiple secret stores each using a different path for their API socket (e.g. https://github.com/crossplane/crossplane/pull/3029/files#r862208395)? Is there a reason to prefer this wrapper approach?

Reply:
No strong reasons other than thinking we could achieve the same with 1 sidecar vs `n` number of sidecars. However, I would agree this could in practice end up having one monolithic sidecar binary which is not good. 
I will update accordingly.

Reply:
Done: https://github.com/crossplane/crossplane/pull/3029/commits/8120ba623f3902a42440be2ecb78757440475848